resources/js/features/users-edit/UserEdit.form.tsx

```diff
@@ -0,0 +1,161 @@
+import {createMutation, createQuery} from "@tanstack/solid-query";
+import {FelteForm, FelteSubmit} from "components/felte-form";
+import {
+  Checkbox,
+  HideableSection,
+  MODAL_STYLE_PRESETS,
+  Modal as ModalComponent,
+  TextField,
+  getTrimInputHandler,
+} from "components/ui";
+import {QueryBarrier, useLangFunc} from "components/utils";
+import {Admin} from "data-access/memo-api/groups/Admin";
+import {AdminUserResource} from "data-access/memo-api/resources/adminUser.resource";
+import {Api} from "data-access/memo-api/types";
+import {Component, createSignal} from "solid-js";
+import {z} from "zod";
+
+export namespace UserEditForm {
+  export const getSchema = () =>
+    z.object({
+      name: z.string(),
+      email: z.string(),
+      hasEmailVerified: z.boolean(),
+      hasPassword: z.boolean(),
+      password: z.string(),
+    });
+
+  export const getInitialValues = (user: AdminUserResource): Readonly<Input> => ({
+    name: user.name,
+    email: user.email || "",
+    hasEmailVerified: user.hasEmailVerified,
+    hasPassword: user.hasPassword,
+    password: "",
+  });
+
+  export type Input = z.input<ReturnType<typeof getSchema>>;
+  export type Output = z.output<ReturnType<typeof getSchema>>;
+
+  interface FormParams {
+    userId: Api.Id;
+  }
+
+  interface Props extends FormParams {
+    onSuccess?: () => void;
+  }
+
+  export const Component: Component<Props> = (props) => {
+    const t = useLangFunc();
+    const userQuery = createQuery(() => Admin.userQueryOptions(props.userId));
+    const user = () => userQuery.data;
+    const invalidate = Admin.useInvalidator();
+    const mutation = createMutation(() => ({
+      mutationFn: Admin.updateUser,
+      onSuccess() {
+        invalidate.users();
+        props.onSuccess?.();
+      },
+    }));
+
+    return (
+      <QueryBarrier queries={[userQuery]}>
+        <FelteForm
+          id="user_edit"
+          onSubmit={(values) =>
+            mutation.mutateAsync({
+              id: user()!.id,
+              name: values.name,
+              ...(values.email
+                ? {
+                    email: values.email,
+                    hasEmailVerified: values.hasEmailVerified,
+                    ...(values.hasPassword
+                      ? user()?.hasPassword && !values.password
+                        ? // The user had password already and it is not changed.
+                          {}
+                        : // New password or a password change.
+                          {password: values.password, passwordExpireAt: null}
+                      : {password: null}),
+                  }
+                : {email: null}),
+            })
+          }
```

Comment:
wyniósłbym takie coś do osobnej funkcji (generowanie danych do requestu), bo w takiej formie jest mało czytelne

Reply:
No to jakby wyniosłem do onSubmit :) Próbowałem to napisać inaczej, seria if'ów modyfikujących requesta aż będzie gotowy, ale naprawdę to co jest tutaj wygląda lepiej.